### PR TITLE
[SPARK-32518][CORE] CoarseGrainedSchedulerBackend.maxNumConcurrentTasks should consider all kinds of resources

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

1.  Make `CoarseGrainedSchedulerBackend.maxNumConcurrentTasks()` considers all kinds of resources when calculating the max concurrent tasks

2. Refactor `calculateAvailableSlots()` to make it be able to be used for both `CoarseGrainedSchedulerBackend` and `TaskSchedulerImpl`

### Why are the changes needed?

Currently, `CoarseGrainedSchedulerBackend.maxNumConcurrentTasks()` only considers the CPU for the max concurrent tasks. This can cause the application to hang when a barrier stage requires extra custom resources but the cluster doesn't have enough corresponding resources. Because, without the checking for other custom resources in `maxNumConcurrentTasks`, the barrier stage can be submitted to the `TaskSchedulerImpl`. But the `TaskSchedulerImpl` won't launch tasks for the barrier stage due to the insufficient task slots calculated by `TaskSchedulerImpl.calculateAvailableSlots` (which does check all kinds of resources).

The application hang issue can be reproduced by the added unit test.

### Does this PR introduce _any_ user-facing change?

Yes. In case of a barrier stage requires more custom resources than the cluster has, the application can get hang before this PR but can fail due to insufficient resources at the end after this PR.

### How was this patch tested?

Added a unit test.

### CODE REVIEW

**Comment by Ngone51** [comment]
ping @tgravescs @jiangxb1987 @mridulm Please take a look, thanks!

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 1]
```diff
@@ -468,51 +468,6 @@ private[spark] class TaskSchedulerImpl(
...
-    offersForResourceProfile.map { case (o, index) =>
-      val numTasksPerExecCores = availableCpus(index) / taskCpus
-      // if limiting resource is empty then we have no other resources, so it has to be CPU
-      if (limitingResource == ResourceProfile.CPUS || limitingResource.isEmpty) {
-        numTasksPerExecCores
-      } else {
-        val taskLimit = resourceProfile.taskResources.get(limitingResource).map(_.amount)
-          .getOrElse {
-            val errorMsg = "limitingResource returns from ResourceProfile " +
-              s"$resourceProfile doesn't actually contain that task resource!"
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:496]: @tgravescs  This actually should not happen, right? According to: 

https://github.com/apache/spark/blob/998086c9a179692b2687bc9a104dbbb35f5a44e2/core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala#L194-L197

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 2]
```diff
@@ -633,9 +633,16 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
   override def maxNumConcurrentTasks(rp: ResourceProfile): Int = synchronized {
-    val cpusPerTask = ResourceProfile.getTaskCpusOrDefaultForProfile(rp, conf)
-    val executorsWithResourceProfile = executorDataMap.values.filter(_.resourceProfileId == rp.id)
-    executorsWithResourceProfile.map(_.totalCores / cpusPerTask).sum
+    val (rpIds, cpus, resources) = executorDataMap.values.toArray.map { executor =>
+      (
+        executor.resourceProfileId,
+        executor.totalCores,
+        executor.resourcesInfo.map { case (name, rInfo) =>
+          (name, rInfo.resourceAddresses.length * rInfo.slotsPerAddress)
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:641]: To be clear, here we are trying to calculate the total concurrent tasks that could run, even if some of them are busy at the moment so I  think we should add some documentation there just to clarify.
Also can we add in a function to ExecutorResourceInfo to get this (rInfo.resourceAddresses.length * rInfo.slotsPerAddress).

**Review by tgravescs — COMMENTED** [review]
ah thanks for fixing, I think this got lost in the change that cores no longer had to be limiting resource and thinking barrier scheduling didn't work with dynamic allocation. I think I was thinking it would be covered when dynamic allocation support was added, but it obviously affects the default profile.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 3]
```diff
@@ -633,9 +633,16 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   }
 
   override def maxNumConcurrentTasks(rp: ResourceProfile): Int = synchronized {
-    val cpusPerTask = ResourceProfile.getTaskCpusOrDefaultForProfile(rp, conf)
-    val executorsWithResourceProfile = executorDataMap.values.filter(_.resourceProfileId == rp.id)
-    executorsWithResourceProfile.map(_.totalCores / cpusPerTask).sum
+    val (rpIds, cpus, resources) = executorDataMap.values.toArray.map { executor =>
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: make sense.

**Review by tgravescs — APPROVED** [review]
lgtm, pending Jenkins

**Comment by tgravescs** [comment]
test this please

**Review by tgravescs — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by Ngone51** [comment]
thanks all! BTW, shall we backport this to branch-3.0? Though, it would be a little different fix since branch-3.0 doesn't have `ResourceProfile`.

**Comment by cloud-fan** [comment]
Yea let's create a new a PR for 3.0

**Comment by tgravescs** [comment]
branch 3 doesn't need it because cores are forced to be limiting resource.

**Comment by Ngone51** [comment]
I think that's the problem. Think about the case where a barrier stage requires 2 CPUs and 2 GPUs but the cluster only has 2 CPUs and 1 GPU. In 3.0, since the limiting resource is cores/CPUs, the barrier stage would have a chance to launch tasks. However, it could only launch one task because the real limiting resources should be GPU. In this case, the barrier stage will fail because of partial task launch. But the error message is quite confusing for users as it suggests to disable delay scheduling, while the real cause should be insufficient (custom) resources. If we backport this fix to 3.0, the barrier stage should fail early before it was able to launch the task.

**Comment by tgravescs** [comment]
This change it checking to see what your executors have. Spark will not let you start the application if your task requirements are such that cpus is not your limiting resource.  so the only way you can get an executor with 2 cores and 1 gpu is if each task requires 2 cores and 1 gpu and at that point gpu is not your limiting resource, they are equivalent.

**Comment by Ngone51** [comment]
That's not true when we don't need to check cores in some clusters? for example, Standalone.

**Comment by tgravescs** [comment]
Oh right, if you didn't specify the executor cores in standalone mode - you get them all by default. Which honestly causes lots of issues - I filed a jira for this for other things. that is left up to the user to properly configure their cluster and job.
There are other things in spark 3.0 that expect cores to be the limiting resource. Dynamic allocation is a big one, which I know you don't care about, but I'm pretty sure there are other things throughout the code that does as well.

We should remove the isDynamicAllocationCheck here so that the warning gets printed for standalone as well:
 https://github.com/apache/spark/blob/branch-3.0/core/src/main/scala/org/apache/spark/SparkContext.scala#L2836

If you want to go through and make sure everything is updated to schedule based on resources, I'm fine with it but I'm pretty sure needs to be more then this.

**Comment by Ngone51** [comment]
I am not sure about the other things(probably, you could link the specific JIRA?) At least, I'd like to backport the fix to schedule based on the resources: https://github.com/apache/spark/pull/29395.

**Pull request closed without merging**